clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java

```diff
@@ -3437,21 +3443,318 @@ public void testPollWithRedundantCreateFetchRequests() {
+        assertThrows(IllegalStateException.class, () -> subscriptions.position(node0Partition2));
+
+        // sendFetches() call #2 should issue a request to node 0 because the first partition in node 0 was collected
+        // (and its buffer removed) and the second partition for node 0 was unassigned. As a result, there are now no
+        // *assigned* partitions for node 0 that are buffered.
+        assertEquals(1, sendFetches());
+    }
+
+    @Test
+    public void testFetchRequestWithBufferedPartitionMissingLeader() {
+        buildFetcher();
+
+        // The test requires that there are multiple nodes as the fetch request logic is based in part off of the
+        // partition-to-node relationship.
+        int numNodes = 2;
+        Set<TopicPartition> partitions = Set.of(tp0, tp1, tp2, tp3);
+        assignFromUser(partitions, numNodes);
+
+        // Get all the nodes serving as the leader for these partitions, and then split them into the separate
+        // nodes and set of partitions to make things easier to keep track of later.
+        List<Node> nodes = nodesForPartitionLeaders(partitions);
+        Node node0 = nodes.get(0);
+        Node node1 = nodes.get(1);
+        List<TopicPartition> node0Partitions = partitionsForNode(node0, partitions);
+        List<TopicPartition> node1Partitions = partitionsForNode(node1, partitions);
+
+        // Seek each partition so that it becomes eligible to fetch.
+        partitions.forEach(tp -> subscriptions.seek(tp, 0));
+
+        // sendFetches() call #1 should issue requests to node 0 or node 1 since neither has buffered data.
+        assertEquals(2, sendFetches());
+        prepareFetchResponses(node0, node0Partitions, 0);
+        prepareFetchResponses(node1, node1Partitions, 0);
+        networkClientDelegate.poll(time.timer(0));
+
+        // Grab both partitions for node 0. The first partition will be collected so that it doesn't have anything
+        // in the fetch buffer. The second node will be left in the buffer, but will clear out its position's leader
+        // node.
+        TopicPartition node0Partition1 = node0Partitions.remove(0);
+        TopicPartition node0Partition2 = node0Partitions.remove(0);
+
+        assertEquals(4, fetcher.fetchBuffer.bufferedPartitions().size());
+        collectSelectedPartition(node0Partition1, partitions);
+        assertEquals(3, fetcher.fetchBuffer.bufferedPartitions().size());
+
+        // Overwrite the position with an empty leader to trigger the test case.
+        SubscriptionState.FetchPosition leaderlessPosition = new SubscriptionState.FetchPosition(
+            0,
+            Optional.empty(),
+            Metadata.LeaderAndEpoch.noLeaderOrEpoch()
+        );
+        subscriptions.position(node0Partition2, leaderlessPosition);
+
+        // Both the collected partition and the position without a partition leader should have a retrievable position.
```

Comment:
Both the collected partition and the position without a partition leader should have a retrievable position. => Both collected partitions should have a retrievable position ?

Reply:
My apologies for the confusing wording. I've revised most of the comments in that method in an effort to improve clarity. PTAL. Thanks.